am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/FilterService.java

```diff
@@ -176,35 +203,36 @@ private void retainFieldsWithReadPermission(JsonNode resource, List<JsonPointer>
+    private void removeFieldsFromData(JsonNode resource, List<JsonPointer> nodesToKeep, List<JsonPointer> nodesToRemove) {
+        nodesToRemove.forEach(pointerCandidateForRemoval -> {
             log.debug(">> Pointer candidate for removal: " + pointerCandidateForRemoval);
-            List<JsonPointer> childPointersWithRead = nodesWithRead.stream()
-                .filter(pointerWithRead -> pointerWithRead.toString().startsWith(pointerCandidateForRemoval.toString() + "/"))
+            List<JsonPointer> visibleChildPointers = nodesToKeep.stream()
+                .filter(visiblePointer -> visiblePointer.toString().startsWith(pointerCandidateForRemoval.toString() + "/"))
                 .collect(Collectors.toList());
-            if (childPointersWithRead.isEmpty()) {
+            if (visibleChildPointers.isEmpty()) {
                 // remove whole node
                 log.debug(">>> Removing '" + pointerCandidateForRemoval + "'");
                 JsonNode node = resource.at(pointerCandidateForRemoval.head());
                 if (node instanceof ObjectNode) {
                     ((ObjectNode) node).remove(pointerCandidateForRemoval.last().toString().substring(1));
                 }
             } else {
-                // retain node's children with READ
+                // retain node's visible children
                 JsonNode node = resource.at(pointerCandidateForRemoval);
                 if (node instanceof ObjectNode) {
-                    childPointersWithRead.forEach(childPointerWithRead -> {
-                        JsonPointer fieldPointer = childPointerWithRead.last();
-                        JsonPointer parentPointer = childPointerWithRead.head();
-
-                        while (!Objects.equals(parentPointer, pointerCandidateForRemoval.head())) {
-                            log.debug(">>> Retaining '" + fieldPointer + "' out of '" + parentPointer + "'");
-                            ((ObjectNode) resource.at(parentPointer)).retain(fieldPointer.toString().substring(1));
-
+                    Map<JsonPointer, Set<String>> visibleNodes = new ConcurrentHashMap<>();
+                    visibleChildPointers.forEach(visibleChildPointer -> {
+                        JsonPointer fieldPointer;
+                        JsonPointer parentPointer = visibleChildPointer;
+                        do {
                             fieldPointer = parentPointer.last();
                             parentPointer = parentPointer.head();
-                        }
+                            visibleNodes.computeIfAbsent(parentPointer, k -> new HashSet<>());
+                            visibleNodes.get(parentPointer).add(fieldPointer.toString().substring(1));
+                        } while (!Objects.equals(pointerCandidateForRemoval, parentPointer));
                     });
+                    visibleNodes.forEach((pointer, fields) -> ((ObjectNode) resource.at(pointer)).retain(fields));
                 }
             }
```

Comment:
can you add java doc comments and explain business logic in brief in the coments? This is very important for code maintenance
It may also be of help to add  audit logs to the methods where business logic is implemented and the method changes the state of the returned output

Reply:
👍 added a Javadoc to this method, and debug logs to the code